### PR TITLE
llama: avoid copying logits during prompt decode in MTP

### PR DESCRIPTION
**Pull request by am17an**
## Overview

Avoid copying the logits for every token in the batch when doing prompt processing for MTP since it only requires the pre-norm. This reduces memory traffic quite a bit and in turn increases PP speed with MTP. 

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure:  YES, for debugging and reviewing

### CODE REVIEW

File: src/llama-context.cpp [hunk 1]
```diff
@@ -1999,6 +2016,12 @@ uint32_t llama_context::output_reserve(int32_t n_outputs) {
     embd.size          = has_embd          ? n_embd_out*n_outputs_max  : 0;
     embd_pre_norm.size = has_embd_pre_norm ? n_embd*n_outputs_max      : 0;
 
+    if (has_embd_pre_norm && !cparams.embeddings_pre_norm_masked) {
+        // unmasked: pre-norm row exists for every token in the ubatch, not just
```
Comment (ggerganov) [inline src/llama-context.cpp]:
```suggestion
        // unmasked: pre-norm row exists for every token in the batch, not just
```

File: src/models/qwen35moe.cpp [hunk 2]
```diff
@@ -234,6 +234,10 @@ llama_model_qwen35moe::graph::graph(const llama_model & model, const llm_graph_p
     cb(cur, "h_pre_norm", -1);
     res->t_h_pre_norm = cur;
```
Comment (ggerganov) [inline src/models/qwen35moe.cpp:235]: Currently, we don't assign this tensor as "output tensor" via the `ggml_set_output()`. This is technically not correct, though in practice since it is near the end of the graph, it is not causing problems.

We have to fix this by adding the corresponding logic to call `ggml_set_output()` in the `llm_graph_result::set_outputs()`.

Comment (am17an) [reply src/models/qwen35moe.cpp:235]: Nice catch! would have been a nasty bug to find

**Review by CISC — APPROVED** [review]

**Review by ggerganov — APPROVED** [review]
A quick bench on RTX 5090 with `Qwen3.6 27B Q4_K`

Image

**Comment by d-r-e** [comment]
> A quick bench on RTX 5090 with `Qwen3.6 27B Q4_K`
> 
> Image

Are the legend colors swapped?

**Comment by pwilkin** [comment]
@d-r-e no, MTP does negatively impact prompt processing, but under this PR the negative impact is halved.

**Comment by cb88** [comment]
2xMI50 qwen 27b Q4_1 does see some improvement with this PR
MI50 without MTP = 500t/s 
with MTP = 250t/s 
with MTP this PR = 300t/s

**Comment by 0cc4m** [comment]
> @d-r-e no, MTP does negatively impact prompt processing, but under this PR the negative impact is halved.

Why does it affect prompt processing?

**Comment by Mithras** [comment]
Made sure this PR is included and re-tested:
unsloth/Qwen3.6-27B-MTP-GGUF/Qwen3.6-27B-UD-Q5_K_XL.gguf
```
| model     |             test |             t/s |     peak t/s |          ttfr (ms) |       est_ppt (ms) |      e2e_ttft (ms) |
|:----------|-----------------:|----------------:|-------------:|-------------------:|-------------------:|-------------------:|
| qwen36-27 |  pp2048 @ d16384 | 1843.93 ± 14.82 |              |   9098.06 ± 124.65 |   9097.48 ± 124.65 |   9098.06 ± 124.65 |
| qwen36-27 |   tg128 @ d16384 |    74.46 ± 3.24 | 84.00 ± 0.82 |                    |                    |                    |
| qwen36-27 |  pp2048 @ d65536 |  1449.72 ± 9.78 |              |  42344.98 ± 292.27 |  42344.40 ± 292.27 |  42344.98 ± 292.27 |
| qwen36-27 |   tg128 @ d65536 |    61.97 ± 2.35 | 68.33 ± 4.78 |                    |                    |                    |
| qwen36-27 | pp2048 @ d131072 |  1075.30 ± 2.36 |              | 112238.48 ± 281.78 | 112237.90 ± 281.78 | 112238.48 ± 281.78 |
| qwen36-27 |  tg128 @ d131072 |    48.40 ± 2.53 | 55.00 ± 0.00 |                    |                    |                    |
```
pretty much the same as  https://github.com/ggml-org/llama.cpp/pull/22673#issuecomment-4471482105 which probably had the PR already. Still almost 50% pp hit

**Comment by pwilkin** [comment]
> Why does it affect prompt processing?

Due to the embeddings copy, most likely.

**Comment by cb88** [comment]
> > Why does it affect prompt processing?
> 
> Due to the embeddings copy, most likely.

Yes PP slows down as context grows, and is slower from the start also, this is probably something that will need to be fixed at the model architecture like with https://github.com/chiennv2000/orthrus so there are no duplicated memory operations between the draft and main model (or at least they are cached)

**Comment by jtjstock** [comment]
This particular commit seems to regress the acceptance rate, I lose about 5% at n = 4 in coding tasks. Otherwise the prefill improvement is amazing

**Comment by am17an** [comment]
@jtjstock this commit just stops copying stuff we don't ever read, in a way it's a free optimization and it should not affect anything except the prefill

**Comment by jtjstock** [comment]
@am17an  Just did a bunch of runs, one at latest(b9254): draft acceptance = 0.75134 ( 3913 accepted /  5208 generated)
and one with this and a dependent commit reverted: draft acceptance = 0.83896 ( 4376 accepted /  5216 generated)

Same prompt. Same Seed. Temp 0. Same hardware minutes apart. Same compiler. The only oddity seems to be the outputs from the latest(b9254), they are slightly different each time, but highly similar, where as with mainline commits (3e12fbdea + dependent 49c21f97c, included this last one for a clean revert and build) reverted build they are identical each time.

I'm running this on windows, the command used:
llama-server.exe ^
  -m models/localweights/Qwen3.6-35B-A3B-MTP-IQ4_XS-Q8nextn-GGUF/Qwen3.6-35B-A3B-MTP-IQ4_XS-Q8nextn.gguf ^
  --no-mmap --ctx-size 16384 --port 12345 ^
  --flash-attn on --cache-type-k q8_0 --cache-type-v q8_0 ^
  --spec-type draft-mtp --spec-draft-n-max 4 ^
  -np 1 --temp 0 --seed 0 ^
  -sm layer -ngl 999 --tensor-split 20,18 ^
  -t 4 -tb 8 --no-warmup --metrics

The prompt: write fizzbuzz in 16 different programming languages

**Comment by jtjstock** [comment]
@am17an Also just tested the official build of b9254(llama-b9254-bin-win-cuda-13.1-x64), seeing the same variability in the output and similar degraded acceptance rate.

Maybe it's specific to my config? It's running on 2x 5060ti 16GB

Edit: did some more testing. Issue still present without quantized kv, but does disappear when I switch to -sm tensor, output from -sm tensor is internally consistent across runs.

**Comment by jtjstock** [comment]
@am17an I think the issue I see, which still presents at head, is the MTP heads are being fed the raw pre-norm residual, but Qwen seems to do better with the post-norm. Before this PR, Qwen was effectively reading a post-norm because t_h_pre_norm aliased the buffer that the final norm was writing to. So it was better before because of a side effect.

It is a very small change to fix. Move the final build_norm before res->t_h_pre_norm = cur in the main graph, and move res->t_h_pre_norm = cur after the build_norm in the mtp one, in qwen35moe.cpp and qwen35.cpp. Two lines moved in each file.

I could be wrong, but it does result in a noticeable improvement to acceptance with the corresponding bump in token generation. I see acceptance move from ~75% to ~80% at n-max 4, and generation from ~203 t/s to ~225 t/s on tensor split. Prefill was of course unaffected.

I'm not looking at the sm layer issue I noted before, as I'm not using layer split anymore, but it's still present when I tested.

**Comment by am17an** [comment]
@jtjstock the model's MTP head was trained on the pre-norm hidden state. I don't understand how the post-norm residual would make it better.

**Comment by jtjstock** [comment]
Try it?

**Comment by am17an** [comment]
No thanks. Please stop posting irrelevant comments

**Comment by jtjstock** [comment]
Sorry I am bothering you, but I don't see any source that specifically says Qwen 3.5/3.6's MTP is trained on pre-norm hidden state, I do see a paper that says the training method is undisclosed(https://arxiv.org/html/2605.09992v1), and vLLM seems to use the post-norm hidden state in the branch qwen 3.5's mtp follows.

Again, I do apologise for bothering you, you've done a tremendous amount of work on this and it is not my intention to waste your time.

**Comment by am17an** [comment]
Okay I just checked vLLM and you're right, they do pass in the post-norm hidden state, my assumption was that is was a deepseek like MTP which passes the pre norm hidden state. Thanks for pointing this out
